feat(premium): fetch user status [IN-979]

cyndichin opened the pull request:
## Summary
Fetch user premium status on sign-in by using `isPremium` field from GraphQL 

## References 
IN-979

## Implementation Details
* Added `isPremium` field to `userByToken` query that gets called when we are Fetching Saves
* Created new class `User` which stores the users status via @AppStorage
* Pass in the `user` to our `FetchList` in order to `setPremiumStatus`
* Add tests and create mocks for user class

## Test Steps
- [ ] WHEN I sign into the app,
THEN request my premium status
THEN store the premium status

- [ ] WHEN the application is launched 
AND a sync request is made (including for user’s premium status)
THEN store the premium status

- [ ] WHEN the user quits the application after launching
THEN the premium status should persist without another request

## PR Checklist:
- [x] Added Unit / UI tests
- [x] Self Review (review, clean up, documentation, run tests)
- [x] Basic Self QA

cyndichin commented on PocketKit/Sources/PocketKit/Home/HomeViewModel.swift:
```diff
@@ -131,7 +131,7 @@ class HomeViewModel {
         networkPathMonitor.updateHandler = { [weak self] path in
             if path.status == .satisfied {
-                self?.refresh { }
+                self?.refresh(isForced: true) { }
```
I missed this during the home refresh work - I believe this should be a force refresh since it should only occur when network connectivity change

cyndichin commented on PocketKit/Sources/Sync/Operations/FetchList.swift:
```diff
@@ -95,7 +100,7 @@ class FetchList: SyncOperation {
     private func fetchPage(_ pagination: PaginationSpec) async throws -> GraphQLResult<FetchSavesQuery.Data> {
         let query = FetchSavesQuery(
-            token: token,
+            token: appSession?.session?.accessToken ?? "",
```
at this point the token should have a value but because we're passing in the appSession and not token, have to do this check here - let me know if there's a better way!

dskuza commented on PocketKit/Sources/SharedPocketKit/Session/AppSession.swift:
```diff
@@ -1,10 +1,21 @@
 public class AppSession {
     @KeychainStorage
     public var currentSession: Session?
+    public var currentStatus: UserStatus?
```
Can we just default to `free` and update as necessary?

cyndichin replied:
I wanted to `nil` it out when the user is logged out, but I can have it default to `free`

dskuza replied:
@cyndichin makes sense; similar thought process to `currentSession` then. an alternative is adding another `loggedOut` case, but I'll leave that decision up to you.

cyndichin replied:
I decided to make an `unknown` case just since the user may be logged in but the Fetch Saves query was not completed.

dskuza reviewed (commented):
@cyndichin Quick question - it doesn't look like we're persisting the user status. Will we lose it if the app is quit until next time a fetch is made? I want to give it another pass, but this stuck out in my mind immediately.

cyndichin commented:
> @cyndichin Quick question - it doesn't look like we're persisting the user status. Will we lose it if the app is quit until next time a fetch is made? I want to give it another pass, but this stuck out in my mind immediately.

Oh yes, you're right. Were you thinking persistence through user defaults? Also, using the app session is a temporary solution for now right, do you foresee using a User object with regards to the Premium work?

dskuza commented:
> Oh yes, you're right. Were you thinking persistence through user defaults? Also, using the app session is a temporary solution for now right, do you foresee using a User object with regards to the Premium work?

Well, we _could_ use `@AppStorage`, or alternatively, we can also add it to `session`, and have the concept of a "free" session or a "premium" session, but that does kind of take away from the concept of a User.

We could also just introduce the concept of a User now. 

Either way, I think the next step before having this pull request fully ready for review is adding persistence to the user status!

cyndichin commented:
> > Oh yes, you're right. Were you thinking persistence through user defaults? Also, using the app session is a temporary solution for now right, do you foresee using a User object with regards to the Premium work?
> 
> Well, we _could_ use `@AppStorage`, or alternatively, we can also add it to `session`, and have the concept of a "free" session or a "premium" session, but that does kind of take away from the concept of a User.
> 
> We could also just introduce the concept of a User now.
> 
> Either way, I think the next step before having this pull request fully ready for review is adding persistence to the user status!

Thanks! I do like the concept of a User and that it's separate from the session. I updated the PR and squashed commits since its pretty different than what I had originally. Will put my PR in ready for review after I rebase the GraphQL issue PR.

cyndichin reviewed (commented):
Curious to know why this `plist` file got generated? 
`PocketKit/Tests/SharedPocketKitTests/UserTests.swift.plist`. I removed it, but was just wondering why it appeared and if anyone faced this when creating a new class.

![image](https://user-images.githubusercontent.com/6743397/200948059-0f9370fa-fa10-463a-9d34-11f1d4964c0c.png)

dskuza reviewed (approved):
😎

cyndichin commented:
> Curious to know why this `plist` file got generated? `PocketKit/Tests/SharedPocketKitTests/UserTests.swift.plist`. I removed it, but was just wondering why it appeared and if anyone faced this when creating a new class.
> 
> ![image](https://user-images.githubusercontent.com/6743397/200948059-0f9370fa-fa10-463a-9d34-11f1d4964c0c.png)

This is because I was using #file instead of a string.

dskuza commented on PocketKit/Sources/SharedPocketKit/User.swift:
```diff
@@ -0,0 +1,32 @@
+    case unknown = "unknown"
+}
+
+public protocol UserProtocol {
```
Basically, and just more readable (to me). It kind of feels like the equivalent of doing something like 

`class ViewModelClass { }`

which just doesn't seem right. 

We know it's a protocol because it's defined as such, so there's no need to add the suffix. That's my thinking, at least.

cyndichin replied:
makes sense, thanks! will do this going forward
